llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp

```diff
@@ -2506,78 +2506,111 @@ bool SelectionDAG::expandFSINCOS(SDNode *Node,
-    Entry.Ty = PointerType::getUnqual(Ty->getContext());
+    Entry.Ty = PointerType::getUnqual(Ctx);
+    ResultPtrs[ResNo] = ResultPtr;
     Args.push_back(Entry);
-    ResultPtrs.push_back(ResultPtr);
   }
 
   SDLoc DL(Node);
 
+  // Pass the vector mask (if required).
   if (VD && VD->isMasked()) {
-    EVT MaskVT = TLI->getSetCCResultType(getDataLayout(), *Ctx, VT);
+    EVT MaskVT = TLI->getSetCCResultType(getDataLayout(), Ctx, VT);
     Entry.Node = getBoolConstant(true, DL, MaskVT, VT);
-    Entry.Ty = MaskVT.getTypeForEVT(*Ctx);
+    Entry.Ty = MaskVT.getTypeForEVT(Ctx);
     Args.push_back(Entry);
   }
 
+  Type *RetType = CallRetResNo.has_value()
+                      ? Node->getValueType(*CallRetResNo).getTypeForEVT(Ctx)
+                      : Type::getVoidTy(Ctx);
   SDValue InChain = StoresInChain ? StoresInChain : getEntryNode();
   SDValue Callee = getExternalSymbol(VD ? VD->getVectorFnName().data() : LCName,
                                      TLI->getPointerTy(getDataLayout()));
   TargetLowering::CallLoweringInfo CLI(*this);
   CLI.setDebugLoc(DL).setChain(InChain).setLibCallee(
-      TLI->getLibcallCallingConv(LC), Type::getVoidTy(*Ctx), Callee,
-      std::move(Args));
+      TLI->getLibcallCallingConv(LC), RetType, Callee, std::move(Args));
 
-  auto [Call, OutChain] = TLI->LowerCallTo(CLI);
+  auto [Call, CallChain] = TLI->LowerCallTo(CLI);
 
   for (auto [ResNo, ResultPtr] : llvm::enumerate(ResultPtrs)) {
+    if (ResNo == CallRetResNo) {
+      Results.push_back(Call);
+      continue;
+    }
     MachinePointerInfo PtrInfo;
     if (StoreSDNode *ST = ResultStores[ResNo]) {
       // Replace store with the library call.
-      ReplaceAllUsesOfValueWith(SDValue(ST, 0), OutChain);
+      ReplaceAllUsesOfValueWith(SDValue(ST, 0), CallChain);
       PtrInfo = ST->getPointerInfo();
     } else {
       PtrInfo = MachinePointerInfo::getFixedStack(
           getMachineFunction(), cast<FrameIndexSDNode>(ResultPtr)->getIndex());
     }
-    SDValue LoadResult = getLoad(VT, DL, OutChain, ResultPtr, PtrInfo);
+    SDValue LoadResult =
+        getLoad(Node->getValueType(ResNo), DL, CallChain, ResultPtr, PtrInfo);
     Results.push_back(LoadResult);
   }
 
+  // FIXME: Find a way to avoid updating the root. This is needed for x86, which
+  // uses a floating-point stack. If (for example) the node to be expanded has
+  // two results one floating-point which is returned by the call, and one
+  // integer result, set returned via an output pointer. If only the integer
+  // result is used then the `CopyFromReg` for the FP result may be optimized
+  // out. This prevents an FP stack pop from being emitted for it. Setting the
+  // root like this ensures there will be a use of the `CopyFromReg` chain, and
+  // ensures the FP pop will be emitted.
+  SDValue OutputChain =
+      getNode(ISD::TokenFactor, DL, MVT::Other, getRoot(), CallChain);
+  setRoot(LegalizeOp ? LegalizeOp(OutputChain) : OutputChain);
```

Comment:
I do not understand this, and the explanation sounds like it's working around something else broken. There shouldn't be any need for a callback. The chain logic should be unconditional, both of these cases have the same properties

Reply:
This workaround was added for the `frexp` expansion in https://github.com/llvm/llvm-project/commit/1d92b68ead0c57a8218c95acd39a2737016ea0ad. I added the `FIXME` as I think there should be a better solution here, but I could not come up with anything decent :sweat_smile: 

The result chain from a `frexp` expansion is something like `f80,ch,glue = CopyFromReg t23, Register:f80 $fp0` (for x86). That corresponds to the FP result, and the integer result comes from an output pointer. 

For the broken cases (fixed in that commit), the FP result is unused. The load of the integer result takes its in-chain from the `CopyFromReg` (which is good). However, in DAG combine, it'll find a better chain that does not include the `CopyFromReg` (it seems it can always skip a `CopyFromReg` to find better memory chains).

That `CopyFromReg` then becomes unused and DCE'd. The x86 backend then won't emit an FP stack pop (`fstp %st(0))`) for the FP result, which is bad...

Updating the root creates a fake use of the `CopyFromReg` chain, which prevents it from being optimized out (and hence keeps the pop). 

This is only really an issue for calls with multiple results (since for single result calls if the result was unused, the node would have been DCD'd before the call expansion)

Reply:
The logic here is unconditional -- the chain is updated the same way callback or not (and I think the legalization is a no-op). However, vector legalization keeps track of the legalized nodes. Since the new root node is not one of the results it does not get automatically legalized, and if not explicitly legalized we'll eventually hit an assert.

Reply:
I realized I can slight tweak the DAG to avoid the need for the callback, so I've done that now :slightly_smiling_face: